cmd/consensus/main.go

```diff
@@ -644,6 +597,97 @@ func main() {
 				return nil, fmt.Errorf("could not find latest finalized block and pending blocks: %w", err)
 			}
 
+			hotstuffModules = &consensus.HotstuffModules{
+				Forks:                nil,
+				Validator:            nil,
+				Notifier:             notifier,
+				Committee:            committee,
+				Signer:               signer,
+				Persist:              persist,
+				Aggregator:           nil,
+				QCCreatedDistributor: pubsub.NewQCCreatedDistributor(),
+			}
```

Comment:
I like this encapsulation of dependencies in `HotstuffModules`, because it makes the method signatures more compact. Specifically for the `HotStuffFactory` for collectors, it   de-bloats the code. 

The specific aspect that concerns me is that we only partially initialize the fields in `HotstuffModules`, while leaving some `nil`. I personally find this somewhat confusing and error prone, because it unnecessarily creates dependencies between different parts of the code, i.e. it reduces modularity. For example: 
* https://github.com/onflow/flow-go/blob/817feda88c1a57fc6c77251e7705f6ff8feaad7f/consensus/participant.go#L143-L148 here, `modules.Forks` is  not nil
* here (same file) `modules.Forks` _is_ nil: https://github.com/onflow/flow-go/blob/817feda88c1a57fc6c77251e7705f6ff8feaad7f/consensus/participant.go#L127
* Hence, I always have to refer back to `consensus.main` and `collection.main` to determine whether I can use a field in `HotstuffModules` to initialize a dependency or whether the field is still nil. Therefore, I feel this 
   * unnecessarily couples the code
   * circumvents helpful compile-time checks, e.g. here: https://github.com/onflow/flow-go/blob/817feda88c1a57fc6c77251e7705f6ff8feaad7f/cmd/consensus/main.go#L611-L623 If I switched the initialization of `hfForks` and the `validator` the compiler will be happy, but the code will crash at run-time, because the reference to Forks will be `nil`. This is only possible, because `HotstuffModules` contains the field `Forks` but we don't initialize it.  
      When creating a reference only when I actually create the instance, I would get a compile-time error about an unknown variable. I consider the latter much safer.  


While I generally like compact code, I would suggest to prioritize API safety higher (even at the cost of more verbose code). Generally (especially in Go), I would suggest to stick with the rule: "Any struct that crosses API boundaries, should contain no `nil` values, unless `nil` is a valid value." In our case, `'nil` is not a valid value for `Forks`, `Validator`, etc. 
  
### Suggestion
* For most of the factory methods in [package `consensus`](https://github.com/onflow/flow-go/blob/817feda88c1a57fc6c77251e7705f6ff8feaad7f/consensus/participant.go) (like `NewValidator`, `NewForks`, `NewVoteAggregator`, `NewFollower`), I would stick with using basic dependencies. I think the overhead is just one or two extra fields in the method signature, which I think is fine.
* However, for `NewParticipant`, we could use  `HotstuffModules`:
   * `NewParticipant` takes a lot of dependencies, so encapsulating some of them substantially reduces the method signature
   * right before we call `NewParticipant`, all fields in `HotstuffModules` are known. I.e. we can create `HotstuffModules`, without any `nil` values. 

I have implemented this suggestion here: https://github.com/onflow/flow-go/pull/1700 (PR is targeting this branch). Some of the tests still need to be updated.

Reply:
Agree, that's kinda what made me feel uneasy as well.